### PR TITLE
Improve entity editor - 1

### PR DESCRIPTION
**Pull request by yusufkinatas**
## 🌟 What is the purpose of this PR?

Follows the introduction of the entity editor in [#1141](https://github.com/hashintel/hash/pull/1141). This PR adds more functionality and improves the code structure around the entity editor.

⚠️ Note that this is still WIP, and there will be follow-up PRs to this PR (e.g. #1121). Various bits of designed functionality remain unimplemented.

## 🔍 What does this change?

- Create links table
- Add sorting to tables, made generic sorting utilities (for now, we only sort by comparing strings)
- Split `property-table` into smaller chunks
- Implement `useSnackbar` to give feedback to user when an update fails
- Create `GlideGrid` component, so we can reuse basic parts of the grid like `theme`
  - Moved some grid related utility functions into `packages/hash/frontend/src/components/GlideGlid/utils.ts`, and made them reuseable for the sake of simplicity for now (this could change in following PR's, especially when we advance on cell rendering more)
- Make entity editor block-like by making it independent from app-specific variables like `route` as @CiaranMn suggested [here](https://github.com/hashintel/hash/pull/1141#discussion_r996858232).

## 🐾 Next steps

- Follow up with more PRs to keep adding missing functionalities, based on the design

## ❓ How to test this?

1. Checkout the branch / view the deployment
4. Login
5. Open /entity-editor url, and copy the entityId of logged in user there. It should be sth like a0718f14-5806-458b-825a-cc00f8b92b8f
6. Open /@username/entities/{entityId} url by using the entityId you just copied
a. To test `links` table, you need to run backend integration tests, and then check local `postgresql` to get some `entityId`'s of entities including links
b. To test different data types on properties table (check the screenshot below), you need to create an entity-type, then an entity with that entity type, which has those data types included. 

## 📹 Demo

Screen Shot 2022-10-18 at 12 10 16

Screen Shot 2022-10-18 at 12 10 01

### CODE REVIEW

File: packages/hash/frontend/src/components/hooks/useSnackbar.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
+import {
+  OptionsObject,
+  ProviderContext,
+  SnackbarKey,
+  SnackbarMessage,
+  // eslint-disable-next-line no-restricted-imports
+  useSnackbar as libUseStackbar,
```
Comment (nathggns) [inline packages/hash/frontend/src/components/hooks/useSnackbar.ts:7]:
```suggestion
  useSnackbar as useLibStackbar,
```

File: packages/hash/frontend/src/components/hooks/useSnackbar.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+  /** we use entries to generate the object with `EnqueueWithoutVariant` function of each variant
+   * so it's really easy-to-use  */
+  return Object.fromEntries(entries) as SnackbarVariants;
+};
+
+export const useSnackbar = () => {
+  const { enqueueSnackbar } = libUseStackbar();
+
+  return generateSnackbarVariants(enqueueSnackbar);
```
Comment (nathggns) [inline packages/hash/frontend/src/components/hooks/useSnackbar.ts:49]: Should this be memo'd?

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/link-table/use-get-cell-content.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+      if (!link) {
+        throw new Error();
+      }
+
+      const linkKey = linkGridIndexes[col];
+
+      if (!linkKey) {
+        throw new Error();
+      }
```
Comment (nathggns) [inline packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/link-table/use-get-cell-content.ts:22]: Should these errors have some kind of description?

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/link-table/use-get-cell-content.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+      return {
+        kind: GridCellKind.Text,
+        data: value,
+        displayData: value,
+        allowOverlay: false,
+      };
+    },
+    /**
+     * @todo check why grid is not updating without adding `linkSort` as dependency
+     * rowData is already depending on `linkSort`
```
Comment (nathggns) [inline packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/link-table/use-get-cell-content.ts:35]: Any thoughts on this?

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/property-table/extract-enriched-property-types-from-entity.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
+import { PropertyType } from "@blockprotocol/type-system-web";
+import { capitalize } from "@mui/material";
+import { EntityResponse } from "../../../../../components/hooks/blockProtocolFunctions/knowledge/knowledge-shim";
+import { EnrichedPropertyType } from "./types";
+
+const getDataTypesOfPropertyType = (
+  propertyType: PropertyType,
+  entity: EntityResponse,
+) => {
+  /** @todo check why propertyValue does not have with a proper type  */
```
Comment (nathggns) [inline packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/property-table/extract-enriched-property-types-from-entity.ts:10]: Any ideas on this?

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/property-table/use-get-cell-content.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+      const property = rowData[row];
+
+      if (!property) {
+        throw new Error();
+      }
+
+      const propertyKey = propertyGridIndexes[col];
+
+      if (!propertyKey) {
+        throw new Error();
```
Comment (nathggns) [inline packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/property-table/use-get-cell-content.ts:21]: Do these need descriptions?

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/property-table/use-on-cell-edited.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+      const updatedProperties = {
+        ...entity.properties,
+        [property.propertyTypeId]: newValue.data,
+      };
+
+      /**
+       * setting state for optimistic update
+       * also storing previous entity, so we can rollback if API call fails
+       */
+      const prevEntity = { ...entity };
```
Comment (nathggns) [inline packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/property-table/use-on-cell-edited.ts:36]: Why do you need to clone it?

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/property-table/use-get-cell-content.ts (hunk 6 above)
Comment (yusufkinatas) [reply packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/property-table/use-get-cell-content.ts:21]: If this error pops up, it means we're using glide grid horribly wrong. Added this to satisfy typescript for the rest of the scope

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/property-table/use-on-cell-edited.ts (hunk 7 above)
Comment (yusufkinatas) [reply packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/property-table/use-on-cell-edited.ts:36]: Hmm, you're right, calling setEntity on the line below does not change the value of `entity` immediately in this scope. I'll remove the clone, but I'll keep `prevEntity`, so it's more clear that we'Re rolling back to previous version if optimistic update fails

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/link-table/use-get-cell-content.ts (hunk 3 above)
Comment (yusufkinatas) [reply packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/link-table/use-get-cell-content.ts:22]: Same https://github.com/hashintel/hash/pull/1207#discussion_r999188396

File: packages/hash/frontend/src/components/hooks/useSnackbar.ts (hunk 1 above)
Comment (yusufkinatas) [reply packages/hash/frontend/src/components/hooks/useSnackbar.ts:7]: Good catch, updated

File: packages/hash/frontend/src/components/hooks/useSnackbar.ts (hunk 2 above)
Comment (yusufkinatas) [reply packages/hash/frontend/src/components/hooks/useSnackbar.ts:49]: Yes, updated!

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/property-table/extract-enriched-property-types-from-entity.ts (hunk 5 above)
Comment (yusufkinatas) [reply packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/property-table/extract-enriched-property-types-from-entity.ts:10]: Updated it, it was a union type, so I should've used property key checking to determine which type it actually is.

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/link-table/use-get-cell-content.ts (hunk 4 above)
Comment (yusufkinatas) [reply packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/link-table/use-get-cell-content.ts]: Found the issue and solved it. It was related with object references.

**Comment by yusufkinatas** [comment]
@nathggns Thanks for the review, all comments are addressed at https://github.com/hashintel/hash/pull/1207/commits/aa0280d44f4082bcccc07153242aed26a61fee9b

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/property-table/use-on-cell-edited.ts (hunk 7 above)
Comment (nathggns) [reply packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/property-table/use-on-cell-edited.ts]: Sure, but you can still put a string describing the error in the event it does happen.

**Comment by nathggns** [comment]
Will re-review later today, but added a response about the error descriptions

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/property-table/use-on-cell-edited.ts (hunk 7 above)
Comment (yusufkinatas) [reply packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/property-table/use-on-cell-edited.ts]: I guess you mention the `throw new Error()`s above

Comment (nathggns) [reply packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/property-table/use-on-cell-edited.ts]: Oh sorry this went to the wrong place – damn github app. Sorry!

**Comment by yusufkinatas** [comment]
Cool, added missing error messages here https://github.com/hashintel/hash/pull/1207/commits/00d84bb70a262a032787949d08dbc9f62d7c1718

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/link-table/use-row-data.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+          linkedWith: generateEntityLabel(link.targetEntity),
+          linkId: link.linkTypeId,
+          relationShip: "Outbound",
+          type: linkType?.title ?? "",
+        };
+      }) ?? []
+    );
+  }, [entity]);
+
+  const sortedRowData = sortRowData(rowData, linkSort);
```
Comment (nathggns) [inline packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/link-table/use-row-data.ts:38]: This looks like it ought to be memo'd. It is for property tables, and this ends up in the dependency array of `useCallback` in `useGetCellContent`, so really ought to be stable

**Review by nathggns — CHANGES_REQUESTED** [review]

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/link-table/use-row-data.ts (hunk 8 above)
Comment (yusufkinatas) [reply packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/link-table/use-row-data.ts]: Ugh, thanks for catching this, there are two `sortedRowData`'s for link table & entity table. I forgot to memoize this one 🤦‍♂️

**Review by nathggns — APPROVED** [review]

**Pull request merged**